drivers/flash/nrf_qspi_nor.c

```diff
@@ -0,0 +1,688 @@
+#else
+	initstruct->prot_if.writeoc = NRF_QSPI_WRITEOC_PP;
+#endif
+	initstruct->prot_if.addrmode =
+		qspi_get_address_size(DT_INST_0_NORDIC_QSPI_NOR_ADDRESS_SIZE_32);
+
+	initstruct->prot_if.dpmconfig = false;
+
+	/* Configure physical interface */
+	initstruct->phy_if.sck_freq =
+		get_nrf_qspi_prescaler(DT_INST_0_NORDIC_QSPI_NOR_SCK_FREQUENCY);
+	initstruct->phy_if.sck_delay = DT_INST_0_NORDIC_QSPI_NOR_SCK_DELAY;
+	initstruct->phy_if.spi_mode = qspi_get_mode(DT_INST_0_NORDIC_QSPI_NOR_CPOL,
+						    DT_INST_0_NORDIC_QSPI_NOR_CPHA);
+
+	initstruct->phy_if.dpmen = false;
+}
+
+/* Configures QSPI memory for the transfer */
+int qspi_nrfx_configure(struct device *dev)
+{
+	if (!dev) {
+		return -ENXIO;
+	}
+
+	/* Main config structure */
+	nrfx_qspi_config_t QSPIconfig;
+
+	qspi_fill_init_struct(&QSPIconfig);
+
+	int res = nrfx_qspi_init(&QSPIconfig, qspi_handler, dev);
+
+	/* If quad transfer was chosen - enable it now */
+	if ((qspi_is_used_write_quad_mode(QSPIconfig.prot_if.writeoc)) ||
+	    (qspi_is_used_read_quad_mode(QSPIconfig.prot_if.readoc))) {
+
+		/* WRITE ENABLE has to be sent before QUAR ENABLE */
+		struct qspi_cmd cmd = {
+			.op_code =  SPI_NOR_CMD_WREN
+		};
+
+		if (qspi_send_cmd(dev, &cmd) != 0) {
+			res = -EIO;
+		}
+
+		u8_t tx = QSPI_NOR_QE_BIT;
+
+		const struct qspi_buf tx_buff = {
+			.buf = &tx,
+			.len = sizeof(tx),
+		};
+
+		cmd.op_code = SPI_NOR_CMD_WRSR;
+		cmd.tx_buf = &tx_buff;
+		cmd.rx_buf = NULL;
+
+		if (qspi_send_cmd(dev, &cmd) != 0) {
+			res = -EIO;
+		}
+	}
+	return qspi_get_zephyr_ret_code(res);
+}
+
+/**
+ * @brief Retrieve the Flash JEDEC ID and compare it with the one expected
+ *
+ * @param dev The device structure
+ * @param flash_id The flash info structure which contains the
+ *		  expected JEDEC ID
+ * @return 0 on success, negative errno code otherwise
+ */
+static inline int qspi_nor_read_id(struct device *dev,
+				   const struct qspi_nor_config *const flash_id)
+{
+	u8_t rx_b[QSPI_NOR_MAX_ID_LEN];
+	struct qspi_buf q_rx_buf = {
+		.buf = rx_b,
+		.len = QSPI_NOR_MAX_ID_LEN
+	};
+	struct qspi_cmd cmd = {
+		.op_code = SPI_NOR_CMD_RDID,
+		.rx_buf = &q_rx_buf,
+		.tx_buf = NULL
+	};
+
+	if (qspi_send_cmd(dev, &cmd) != 0) {
+		return -EIO;
+	}
+
+	if (memcmp(flash_id->id, rx_b, QSPI_NOR_MAX_ID_LEN) != 0) {
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static int qspi_nor_read(struct device *dev, off_t addr, void *dest,
+			 size_t size)
+{
+	if (!dest) {
+		return -EINVAL;
+	}
+
+	/* read size must be multiple of 4 bytes */
+	if (size % sizeof(uint32_t) ||
+	    !(size > 0)) {
+		return -EINVAL;
+	}
+
+	qspi_lock(dev);
+
+	const struct qspi_nor_config *params = dev->config->config_info;
+
+	/* should be between 0 and flash size */
+	if ((addr < 0) || ((addr + size) > params->size)) {
+		return -EINVAL;
+	}
+	struct qspi_buf rx_buf = {
+		.buf = dest,
+		.len = size
+	};
+
+	int ret = nrfx_qspi_read(rx_buf.buf, rx_buf.len, addr);
```

Comment:
What's the point in involving `rx_buf` while you could use `dest` and `size` directly here?
Same goes for the call to `nrfx_qspi_write` below.

Reply:
Indeed, it is not needed anymore. Removed it.